File: frontend/src/auth/roles.ts
/**
 * Application roles — aligned with backend RBAC (core/roles.py).
 *
 * Canonical roles are returned by GET /api/me/. Legacy slugs (investor, hr)
 * are still accepted from mock sessions and older rows.
 */
export const ROLES = {
    ADMIN: "admin",
    TENDER: "tender",
    TENDER_USER: "tender_user",
    SUPPLIER: "supplier",
    HR_ADMIN: "hr_admin",
    CANDIDATE: "candidate",
    DESIGNER: "designer",
    /** @deprecated demo alias — maps to tender_user */
    INVESTOR: "investor",
    /** @deprecated legacy — maps to hr_admin */
    HR: "hr",
    /** Legacy interviewer workflows */
    INTERVIEWER: "interviewer",
} as const;

export type AppRole = (typeof ROLES)[keyof typeof ROLES];

/** Roles stored canonically on the backend User model */
export type CanonicalRole =
    | typeof ROLES.ADMIN
    | typeof ROLES.TENDER
    | typeof ROLES.TENDER_USER
    | typeof ROLES.SUPPLIER
    | typeof ROLES.HR_ADMIN
    | typeof ROLES.CANDIDATE
    | typeof ROLES.DESIGNER;

const ALL_ROLES: AppRole[] = Object.values(ROLES);

const ROLE_ALIASES: Record<string, AppRole> = {
    investor: ROLES.TENDER_USER,
    hr: ROLES.HR_ADMIN,
};

/** Human-readable labels for UI (canonical + legacy keys). */
export const ROLE_LABELS: Record<string, string> = {
    admin: "Administrator",
    tender: "Tender",
    tender_user: "Tender user",
    supplier: "Supplier",
    hr_admin: "HR Admin",
    candidate: "Candidate",
    designer: "Designer",
    interviewer: "Interviewer",
    investor: "Tender user",
    hr: "HR Admin",
};

export function normalizeRole(raw: string | null | undefined): AppRole | null {
    if (!raw || typeof raw !== "string") return null;
    const value = raw.trim().toLowerCase();
    if (!value) return null;
    const mapped = (ROLE_ALIASES[value] ?? value) as AppRole;
    if (ALL_ROLES.includes(mapped)) return mapped;
    if (value === ROLES.INTERVIEWER) return ROLES.INTERVIEWER;
    return null;
}

export function roleLabel(raw: string | null | undefined): string {
    const normalized = normalizeRole(raw);
    if (normalized && ROLE_LABELS[normalized]) return ROLE_LABELS[normalized];
    if (raw && ROLE_LABELS[raw]) return ROLE_LABELS[raw];
    return raw?.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase()) ?? "Guest";
}

/** True when user's role matches any allowed entry (aliases respected). */
export function roleInList(userRole: AppRole | null, allowed: AppRole[]): boolean {
    if (!userRole) return false;
    const normalized = normalizeRole(userRole);
    return allowed.some((entry) => {
        const allowedNorm = normalizeRole(entry);
        return entry === userRole || allowedNorm === normalized;
    });
}

export const PROCUREMENT_STAFF_ROLES: AppRole[] = [
    ROLES.ADMIN,
    ROLES.TENDER,
    ROLES.TENDER_USER,
    ROLES.INVESTOR,
];

export const PROCUREMENT_SUPPLIER_ROLES: AppRole[] = [ROLES.SUPPLIER];

export const PROCUREMENT_ROLES: AppRole[] = [
    ...PROCUREMENT_STAFF_ROLES,
    ...PROCUREMENT_SUPPLIER_ROLES,
];

export const HR_STAFF_ROLES: AppRole[] = [ROLES.ADMIN, ROLES.HR_ADMIN, ROLES.HR];

export const HR_INTERVIEWER_ROLES: AppRole[] = [ROLES.ADMIN, ROLES.INTERVIEWER];

export const DESIGN_STAFF_ROLES: AppRole[] = [ROLES.ADMIN, ROLES.DESIGNER];

File: backend/smartflow/core/roles.py
"""
Application roles and helpers for RBAC (Steps 1–3).

`public` is not stored on User — it represents unauthenticated access only.
"""

from __future__ import annotations

from typing import Final

# --- Canonical roles (stored on User.role after normalization) ---
ROLE_ADMIN: Final = "admin"
ROLE_HR_ADMIN: Final = "hr_admin"
ROLE_TENDER_USER: Final = "tender_user"
ROLE_SUPPLIER: Final = "supplier"
ROLE_CANDIDATE: Final = "candidate"

# Legacy values still present in older rows / JWT sessions until migrated
ROLE_HR_LEGACY: Final = "hr"
ROLE_INTERVIEWER_LEGACY: Final = "interviewer"
# Frontend mock/demo alias (not written to User by default)
ROLE_INVESTOR_LEGACY: Final = "investor"

ALL_ROLES: Final = frozenset(
    {
        ROLE_ADMIN,
        ROLE_HR_ADMIN,
        ROLE_TENDER_USER,
        ROLE_SUPPLIER,
        ROLE_CANDIDATE,
        ROLE_HR_LEGACY,
        ROLE_INTERVIEWER_LEGACY,
        ROLE_INVESTOR_LEGACY,
    }
)

ROLE_LABELS: Final = {
    ROLE_ADMIN: "Administrator",
    ROLE_HR_ADMIN: "HR Admin",
    ROLE_TENDER_USER: "Tender user",
    ROLE_SUPPLIER: "Supplier",
    ROLE_CANDIDATE: "Candidate",
    ROLE_HR_LEGACY: "HR (legacy)",
    ROLE_INTERVIEWER_LEGACY: "Interviewer (legacy)",
    ROLE_INVESTOR_LEGACY: "Investor (legacy)",
}

# Roles allowed on self-service registration (immediate account)
REGISTRATION_ROLES: Final = frozenset({ROLE_CANDIDATE})

# Supplier self-registration creates a pending request — admin approval required
SUPPLIER_REGISTRATION_ROLE: Final = ROLE_SUPPLIER

# Map legacy / demo role strings to canonical role for API responses
ROLE_ALIASES: Final = {
    ROLE_HR_LEGACY: ROLE_HR_ADMIN,
    ROLE_INVESTOR_LEGACY: ROLE_TENDER_USER,
}


def normalize_role(raw: str | None) -> str | None:
    """Return canonical role slug or None."""
    if not raw or not isinstance(raw, str):
        return None
    value = raw.strip().lower()
    if not value:
        return None
    return ROLE_ALIASES.get(value, value)


def is_valid_role(raw: str | None) -> bool:
    normalized = normalize_role(raw)
    return normalized in ALL_ROLES if normalized else False


def role_label(raw: str | None) -> str:
    normalized = normalize_role(raw) or raw or ""
    return ROLE_LABELS.get(normalized, normalized.replace("_", " ").title() or "Unknown")


def effective_role(raw: str | None) -> str | None:
    """Normalized role if valid, else None."""
    normalized = normalize_role(raw)
    return normalized if normalized and normalized in ALL_ROLES else None


# --- Role groups for permission checks (Step 3) ---
PROCUREMENT_STAFF_ROLES: Final = frozenset({ROLE_ADMIN, ROLE_TENDER_USER})
PROCUREMENT_SUPPLIER_ROLES: Final = frozenset({ROLE_SUPPLIER})
HR_STAFF_ROLES: Final = frozenset({ROLE_ADMIN, ROLE_HR_ADMIN, ROLE_HR_LEGACY, ROLE_INTERVIEWER_LEGACY})
HR_CANDIDATE_ROLES: Final = frozenset({ROLE_CANDIDATE})


def user_has_role(user, allowed: frozenset[str]) -> bool:
    if not user or not getattr(user, "is_authenticated", False):
        return False
    canonical = effective_role(getattr(user, "role", None))
    raw = getattr(user, "role", None)
    if canonical and canonical in allowed:
        return True
    return bool(raw and raw in allowed)


def is_procurement_staff(user) -> bool:
    return user_has_role(user, PROCUREMENT_STAFF_ROLES)


def is_supplier_user(user) -> bool:
    return user_has_role(user, PROCUREMENT_SUPPLIER_ROLES)


def is_hr_staff(user) -> bool:
    return user_has_role(user, HR_STAFF_ROLES)


def is_candidate_user(user) -> bool:
    return user_has_role(user, HR_CANDIDATE_ROLES)


def is_admin_user(user) -> bool:
    return user_has_role(user, frozenset({ROLE_ADMIN}))
